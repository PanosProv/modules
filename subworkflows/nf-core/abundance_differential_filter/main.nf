//
// Perform differential analysis
//

include { LIMMA_DIFFERENTIAL                  } from '../../../modules/nf-core/limma/differential/main'
include { LIMMA_DIFFERENTIAL as LIMMA_NORM    } from '../../../modules/nf-core/limma/differential/main'
include { DESEQ2_DIFFERENTIAL                 } from '../../../modules/nf-core/deseq2/differential/main'
include { DESEQ2_DIFFERENTIAL as DESEQ2_NORM  } from '../../../modules/nf-core/deseq2/differential/main'
include { CUSTOM_FILTERDIFFERENTIALTABLE      } from '../../../modules/nf-core/custom/filterdifferentialtable/main'

workflow ABUNDANCE_DIFFERENTIAL_FILTER {
    take:
    // Things we may need to iterate
    ch_input                 // [[meta_input], counts, analysis method, fc_threshold, padj_threshold]

    // Workflow-wide things, we don't need to iterate
    ch_samplesheet           // [ meta_exp, samplesheet ]
    ch_transcript_lengths    // [ meta_exp, transcript_lengths]
    ch_control_features      // [meta_exp, control_features]
    ch_contrasts             // [[ meta_contrast, contrast_variable, reference, target ]]

    main:

    // We need to cross the things we're iterating
    inputs = ch_input
        .combine(ch_samplesheet)
        .combine(ch_contrasts)
        .multiMap { meta_input, abundance, analysis_method, fc_threshold, padj_threshold, meta_exp, samplesheet, contrast_meta, variable, reference, target ->
            samples_and_matrix:
                meta_map = meta_input + [ 'method': analysis_method, 'fc': fc_threshold, 'padj': padj_threshold ]
                [meta_map, samplesheet, abundance]
            contrasts:
                meta_map = contrast_meta + [ 'method': analysis_method ]
                [ meta_map, variable, reference, target]

        }

    // Perform normalization and differential analysis
    DESEQ2_NORM(
        inputs.contrasts.filter{it[0].method == 'deseq2'}.first(),
        inputs.samples_and_matrix.filter{it[0].method == 'deseq2'},
        ch_control_features,
        ch_transcript_lengths
    )

    LIMMA_NORM(
        inputs.contrasts.filter{it[0].method == 'limma'}.first(),
        inputs.samples_and_matrix.filter{it[0].method == 'limma'}
    )

    DESEQ2_DIFFERENTIAL(
        inputs.contrasts.filter{it[0].method == 'deseq2'},
        inputs.samples_and_matrix.filter{it[0].method == 'deseq2'},
        ch_control_features,
        ch_transcript_lengths
    )

    LIMMA_DIFFERENTIAL(
        inputs.contrasts.filter{it[0].method == 'limma'},
        inputs.samples_and_matrix.filter { it[0].method == 'limma' }
    )

    // Combine results
    ch_results = DESEQ2_DIFFERENTIAL.out.results.mix(LIMMA_DIFFERENTIAL.out.results)
    ch_normalised_matrix = DESEQ2_NORM.out.normalised_counts.mix(LIMMA_NORM.out.normalised_counts)
    ch_model = DESEQ2_DIFFERENTIAL.out.model.mix(LIMMA_DIFFERENTIAL.out.model)
    ch_versions = DESEQ2_DIFFERENTIAL.out.versions
        .mix(LIMMA_DIFFERENTIAL.out.versions)

    // Extract the fc and pval filters from the metamap we stashed them in
    ch_diff_filter_params = ch_results.multiMap { meta, results ->
        fc_column: meta.method == 'deseq2' ? 'log2FoldChange' : 'logFC'
        padj_column: meta.method == 'deseq2' ? 'padj' : 'adj.P.Val'
        fc_threshold: meta.fc
        padj_threshold: meta.padj
    }

    // Filter differential results
    CUSTOM_FILTERDIFFERENTIALTABLE(
        ch_results,
        ch_diff_filter_params.fc_column,
        ch_diff_filter_params.fc_threshold,
        ch_diff_filter_params.padj_column,
        ch_diff_filter_params.padj_threshold
    )

    emit:
    results_genewise           = ch_results.map{meta, results -> [meta, results] }
    results_genewise_filtered  = CUSTOM_FILTERDIFFERENTIALTABLE.out.filtered
    normalised_matrix          = ch_normalised_matrix
    variance_stabilised_matrix = DESEQ2_NORM.out.rlog_counts.mix(DESEQ2_NORM.out.vst_counts)
    model                      = ch_model
    versions                   = ch_versions
}
