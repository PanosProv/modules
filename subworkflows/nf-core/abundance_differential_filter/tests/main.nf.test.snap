{
    "limma - basic - microarray": {
        "content": [
            [
                [
                    {
                        "id": "diagnosis_normal_uremia_test",
                        "variable": "diagnosis",
                        "reference": "normal",
                        "target": "uremia",
                        "method": "limma"
                    },
                    "diagnosis_normal_uremia_test_limma.limma.results.tsv:md5,4f0944b54b3ab7aa6ded5dd4b4e82802"
                ]
            ],
            [
                [
                    {
                        "id": "diagnosis_normal_uremia_test",
                        "variable": "diagnosis",
                        "reference": "normal",
                        "target": "uremia",
                        "method": "limma",
                        "fc_threshold": 1.5,
                        "padj_threshold": 0.05
                    },
                    "diagnosis_normal_uremia_test_filtered.tsv:md5,a971455ece4ae3c3ab902407b36fc6a5"
                ]
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "diagnosis_normal_uremia_test",
                        "variable": "diagnosis",
                        "reference": "normal",
                        "target": "uremia",
                        "method": "limma"
                    },
                    "diagnosis_normal_uremia_test_limma.limma.model.txt:md5,70b000f632b8bdba4917046362dd876b"
                ]
            ],
            [
                "versions.yml:md5,1ddaab440e2528c688c05a02dd066f12"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-07T14:58:10.009703817"
    },
    "deseq2 and limma - mouse - basic": {
        "content": [
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2.deseq2.results.tsv:md5,791cdba2615a445cded13cae95df73ef"
                ],
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "limma"
                    },
                    "treatment_mCherry_hND6__test_limma_voom.limma.results.tsv:md5,ff36827b7869a8a3c3c905efedcafc93"
                ],
                [
                    {
                        "id": "treatment_mCherry_hND6_sample_number_test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "sample_number",
                        "method": "limma"
                    },
                    "treatment_mCherry_hND6_sample_number_test_limma_voom.limma.results.tsv:md5,ff36827b7869a8a3c3c905efedcafc93"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2",
                        "fc_threshold": 1.5,
                        "padj_threshold": 0.05
                    },
                    "treatment_mCherry_hND6__test_filtered.tsv:md5,7829ead408f4c9cad4277598a231c494"
                ],
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "limma",
                        "fc_threshold": 1.5,
                        "padj_threshold": 0.05
                    },
                    "treatment_mCherry_hND6__test_filtered.tsv:md5,0bfc9215edc6aad064c3ce6abc81bfce"
                ],
                [
                    {
                        "id": "treatment_mCherry_hND6_sample_number_test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "sample_number",
                        "method": "limma",
                        "fc_threshold": 1.5,
                        "padj_threshold": 0.05
                    },
                    "treatment_mCherry_hND6_sample_number_test_filtered.tsv:md5,0bfc9215edc6aad064c3ce6abc81bfce"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2_voom_norm.normalised_counts.tsv:md5,46ab7200c626649ab6256ed797ef5071"
                ],
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "limma"
                    },
                    "treatment_mCherry_hND6__test_limma_voom_norm.normalised_counts.tsv:md5,2aa4880ba5ae246a728b25f4316ca2ca"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2_voom_norm.rlog.tsv:md5,b1adc1fba6bd0c8b55973608f4b97030"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2.deseq2.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                ],
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "limma"
                    },
                    "treatment_mCherry_hND6__test_limma_voom.limma.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                ],
                [
                    {
                        "id": "treatment_mCherry_hND6_sample_number_test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "sample_number",
                        "method": "limma"
                    },
                    "treatment_mCherry_hND6_sample_number_test_limma_voom.limma.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                ]
            ],
            [
                "versions.yml:md5,1ddaab440e2528c688c05a02dd066f12",
                "versions.yml:md5,1ddaab440e2528c688c05a02dd066f12",
                "versions.yml:md5,2c0576aefff8da32c7c0cfd8529aa4b5"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-07T15:00:00.494446257"
    },
    "limma - voom": {
        "content": [
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "limma",
                        "fc_threshold": 1.5,
                        "padj_threshold": 0.05
                    },
                    "treatment_mCherry_hND6__test_filtered.tsv:md5,0bfc9215edc6aad064c3ce6abc81bfce"
                ],
                [
                    {
                        "id": "treatment_mCherry_hND6_sample_number_test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "sample_number",
                        "method": "limma",
                        "fc_threshold": 1.5,
                        "padj_threshold": 0.05
                    },
                    "treatment_mCherry_hND6_sample_number_test_filtered.tsv:md5,0bfc9215edc6aad064c3ce6abc81bfce"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "limma"
                    },
                    "treatment_mCherry_hND6__test_limma_voom_norm.normalised_counts.tsv:md5,2aa4880ba5ae246a728b25f4316ca2ca"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "limma"
                    },
                    "treatment_mCherry_hND6__test_limma_voom.limma.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                ],
                [
                    {
                        "id": "treatment_mCherry_hND6_sample_number_test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "sample_number",
                        "method": "limma"
                    },
                    "treatment_mCherry_hND6_sample_number_test_limma_voom.limma.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                ]
            ],
            [
                "versions.yml:md5,1ddaab440e2528c688c05a02dd066f12",
                "versions.yml:md5,1ddaab440e2528c688c05a02dd066f12"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-07T14:58:27.13269854"
    },
    "stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2"
                        },
                        "treatment_mCherry_hND6__test.deseq2.results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2",
                            "fc_threshold": 1.5,
                            "padj_threshold": 0.05
                        },
                        "treatment_mCherry_hND6__test_filtered.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2"
                        },
                        "treatment_mCherry_hND6__test.normalised_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2"
                        },
                        "treatment_mCherry_hND6__test.rlog.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2"
                        },
                        "treatment_mCherry_hND6__test.deseq2.model.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    "versions.yml:md5,05e3901f6d78f8839a7e07f422e9bc03"
                ],
                "model": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2"
                        },
                        "treatment_mCherry_hND6__test.deseq2.model.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "normalised_matrix": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2"
                        },
                        "treatment_mCherry_hND6__test.normalised_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "results_genewise": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2"
                        },
                        "treatment_mCherry_hND6__test.deseq2.results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "results_genewise_filtered": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2",
                            "fc_threshold": 1.5,
                            "padj_threshold": 0.05
                        },
                        "treatment_mCherry_hND6__test_filtered.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "variance_stabilised_matrix": [
                    [
                        {
                            "id": "treatment_mCherry_hND6__test",
                            "variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking": "",
                            "method": "deseq2"
                        },
                        "treatment_mCherry_hND6__test.rlog.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,05e3901f6d78f8839a7e07f422e9bc03"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-07T15:00:26.610915878"
    },
    "deseq2 - mouse - basic": {
        "content": [
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2.deseq2.results.tsv:md5,791cdba2615a445cded13cae95df73ef"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2",
                        "fc_threshold": 1.5,
                        "padj_threshold": 0.05
                    },
                    "treatment_mCherry_hND6__test_filtered.tsv:md5,7829ead408f4c9cad4277598a231c494"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2_voom_norm.normalised_counts.tsv:md5,46ab7200c626649ab6256ed797ef5071"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2_voom_norm.rlog.tsv:md5,b1adc1fba6bd0c8b55973608f4b97030"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2.deseq2.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                ]
            ],
            [
                "versions.yml:md5,2c0576aefff8da32c7c0cfd8529aa4b5"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-07T14:57:12.932252345"
    },
    "deseq2 - with transcript lengths": {
        "content": [
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2.deseq2.results.tsv:md5,944176b73455aa7c8de3ec32c03edef6"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2",
                        "fc_threshold": 1.5,
                        "padj_threshold": 0.05
                    },
                    "treatment_mCherry_hND6__test_filtered.tsv:md5,a48241fb5f24d961d3cce208f060b624"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2_voom_norm.normalised_counts.tsv:md5,7050f44c460cc13e3f101d048d503527"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2_voom_norm.rlog.tsv:md5,22a4b117246b2317e0f4daf7919703f2"
                ]
            ],
            [
                [
                    {
                        "id": "treatment_mCherry_hND6__test",
                        "variable": "treatment",
                        "reference": "mCherry",
                        "target": "hND6",
                        "blocking": "",
                        "method": "deseq2"
                    },
                    "treatment_mCherry_hND6__test_deseq2.deseq2.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                ]
            ],
            [
                "versions.yml:md5,2c0576aefff8da32c7c0cfd8529aa4b5"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-07T14:59:04.110921498"
    }
}